python/fusion_engine_client/parsers/mixed_log_reader.py

```diff
@@ -0,0 +1,342 @@
+                if self.return_offset:
+                    result.append(offset_bytes)
+                return result
+            except ValueError as e:
+                offset_bytes += 1
+                self.logger.trace('%s Rewinding to offset %d (0x%x).' % (str(e), offset_bytes, offset_bytes),
+                                  depth=2)
+                self.input_file.seek(offset_bytes, os.SEEK_SET)
+
+        # Out of the loop - EOF reached.
+        self._print_progress(self.total_bytes_read)
+        self.logger.debug("Read %d bytes total." % self.total_bytes_read)
+
+        # If we are creating an index file, save it now.
+        if self.index_builder is not None:
+            self.logger.debug("Saving index file as '%s'." % self.index_path)
+            self.index_builder.save(self.index_path, self.input_file.name)
+
+        # Finished iterating.
+        raise StopIteration()
+
+    def _advance_to_next_sync(self):
+        if self.index is None:
+            try:
+                while True:
+                    byte0 = self.input_file.read(1)[0]
+                    while True:
+                        if byte0 == MessageHeader._SYNC0:
+                            byte1 = self.input_file.read(1)[0]
+                            if byte1 == MessageHeader._SYNC1:
+                                self.input_file.seek(-2, os.SEEK_CUR)
+                                return True
+                            byte0 = byte1
+                        else:
+                            break
+            except IndexError:
+                return False
+        else:
+            if self.next_index_elem == len(self.index):
+                return False
+            else:
+                offset_bytes = self.index._data['offset'][self.next_index_elem]
+                self.next_index_elem += 1
+                self.input_file.seek(offset_bytes, os.SEEK_SET)
+                return True
+
+    def _print_progress(self, file_size=None):
+        if file_size is None:
+            file_size = self.file_size_bytes
+
+        if self.total_bytes_read - self.last_print_bytes > 10e6 or self.total_bytes_read == file_size:
+            elapsed_sec = (datetime.now() - self.start_time).total_seconds()
+            self.logger.log(logging.INFO if self.show_progress else logging.DEBUG,
+                            'Processed %d/%d bytes (%.1f%%). [elapsed=%.1f sec, rate=%.1f MB/s]' %
+                            (self.total_bytes_read, file_size,
+                             100.0 * float(self.total_bytes_read) / file_size,
+                             elapsed_sec, self.total_bytes_read / elapsed_sec / 1e6))
+            self.last_print_bytes = self.total_bytes_read
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        return self.next()
+
+    def __getitem__(self, key):
+        """!
+        @brief Limit the returned messages by type or time.
+
+        @warning
+        This operator modifies this class in-place.
```

Comment:
This seems unexpected enough that I would suggest not using the built in `__getitem__` and to use function name that might suggest this behavior more strongly something like `filter_inplace`.

Reply:
Yeah, good call. Ideally this would be able to return a new reader with the additional filter criteria, but since the reader stores a file object internally and there's no trivial way to copy that, we can't really clone the reader. I'll rename this.